src/hb-subset-cff1.cc

```diff
@@ -883,186 +701,204 @@ struct cff_subset_plan {
   bool		desubroutinize;
 };
 
-static inline bool _write_cff1 (hb_serialize_context_t *c,
-				const cff_subset_plan &plan,
-				const OT::cff1::accelerator_subset_t  &acc,
-				unsigned int num_glyphs)
+static bool _serialize_cff1 (hb_serialize_context_t *c,
+			     cff_subset_plan &plan,
+			     const OT::cff1::accelerator_subset_t  &acc,
+			     unsigned int num_glyphs)
 {
-  OT::cff1 *cff = c->allocate_min<OT::cff1> ();
-  if (unlikely (!cff))
-    return false;
+  /* private dicts & local subrs */
+  for (int i = (int)acc.privateDicts.length; --i >= 0 ;)
+  {
+    if (plan.fdmap.has (i))
+    {
+      objidx_t	subrs_link = 0;
+      if (plan.subset_localsubrs[i].length > 0)
+      {
+	CFF1Subrs *dest = c->start_embed <CFF1Subrs> ();
+	if (unlikely (dest == nullptr)) return false;
+	c->push ();
+	if (likely (dest && dest->serialize (c, plan.subset_localsubrs[i])))
+	  subrs_link = c->pop_pack ();
+	else
+	{
+	  c->pop_discard ();
+	  return false;
+	}
+      }
 
-  /* header */
-  cff->version.major = 0x01;
-  cff->version.minor = 0x00;
-  cff->nameIndex = cff->min_size;
-  cff->offSize = 4; /* unused? */
+      PrivateDict  *pd = c->start_embed<PrivateDict> ();
+      if (unlikely (pd == nullptr)) return false;
+      c->push ();
+      cff_private_dict_op_serializer_t privSzr (plan.desubroutinize, plan.drop_hints);
+      /* N.B. local subrs immediately follows its corresponding private dict. i.e., subr offset == private dict size */
+      if (likely (pd->serialize (c, acc.privateDicts[i], privSzr, subrs_link)))
+      {
+	unsigned fd = plan.fdmap[i];
+	plan.fontdicts_mod[fd].privateDictInfo.size = c->length ();
+	plan.fontdicts_mod[fd].privateDictInfo.link = c->pop_pack ();
+      }
+      else
+      {
+	c->pop_discard ();
+	return false;
+      }
+    }
+  }
 
-  /* name INDEX */
+  if (!acc.is_CID ())
+    plan.info.privateDictInfo = plan.fontdicts_mod[0].privateDictInfo;
+
+  /* CharStrings */
   {
-    assert (cff->nameIndex == (unsigned) (c->head - c->start));
-    CFF1NameIndex *dest = c->start_embed<CFF1NameIndex> ();
-    if (unlikely (dest == nullptr)) return false;
-    if (unlikely (!dest->serialize (c, *acc.nameIndex)))
+    CFF1CharStrings  *cs = c->start_embed<CFF1CharStrings> ();
+    if (unlikely (cs == nullptr)) return false;
+    c->push ();
+    if (likely (cs->serialize (c, plan.subset_charstrings)))
+      plan.info.char_strings_link = c->pop_pack ();
+    else
     {
-      DEBUG_MSG (SUBSET, nullptr, "failed to serialize CFF name INDEX");
+      c->pop_discard ();
       return false;
     }
   }
 
-  /* top dict INDEX */
+  /* FDArray (FD Index) */
+  if (acc.fdArray != &Null(CFF1FDArray))
```

Comment:
Can we just check array's count?

Reply:
It didn't occur to me accessing a null obj for non-error cases. Yes, it should work and I can make that change.

Reply:
> It didn't occur to me accessing a null obj for non-error cases.

Yes, we prefer always doing so, handling error just like empty structs.